src/rate_limiter.cc

```diff
@@ -158,19 +158,78 @@ RateLimiter::UnregisterModel(const TritonModel* model)
   }
 }
 
-bool
-RateLimiter::PayloadSlotAvailable(const TritonModel* model)
+void
+RateLimiter::WaitForConsumer(
+    const TritonModel* model, const TritonModelInstance* model_instance)
 {
-  bool result;
   PayloadQueue* payload_queue = nullptr;
   {
     std::lock_guard<std::mutex> lk(payload_queues_mu_);
+    if (payload_queues_.find(model) == payload_queues_.end()) {
+      LOG_ERROR << "Should not print this! Waiting for the consumer for an "
+                   "unknown model";
+      return;
+    }
     payload_queue = payload_queues_[model].get();
   }
+
+  if (model_instance == nullptr) {
+    payload_queue->queue_->WaitForConsumer();
+  } else {
+    payload_queue->specific_queues_[model_instance]->WaitForConsumer();
+  }
+}
+
+
+int
+RateLimiter::WaitingConsumerCount(
+    const TritonModel* model, const TritonModelInstance* model_instance)
+{
+  PayloadQueue* payload_queue = nullptr;
   {
-    std::lock_guard<std::mutex> lk(payload_queue->mu_);
-    result = payload_queue->queue_->Size() <
-             2 * payload_queue->specific_queues_.size();
+    std::lock_guard<std::mutex> lk(payload_queues_mu_);
+    if (payload_queues_.find(model) == payload_queues_.end()) {
+      LOG_ERROR << "Should not print this! Waiting for the consumer for an "
+                   "unknown model";
+      return 0;
+    }
+    payload_queue = payload_queues_[model].get();
+  }
+
+  if (model_instance == nullptr) {
+    return payload_queue->queue_->WaitingConsumerCount();
+  } else {
+    return payload_queue->specific_queues_[model_instance]
+        ->WaitingConsumerCount();
+  }
+}
+
+bool
+RateLimiter::PayloadSlotAvailable(
+    const TritonModel* model, const TritonModelInstance* model_instance,
+    const bool support_prefetching, const bool force_non_blocking)
+{
+  bool result;
+  if (support_prefetching) {
+    PayloadQueue* payload_queue = nullptr;
+    {
+      std::lock_guard<std::mutex> lk(payload_queues_mu_);
+      payload_queue = payload_queues_[model].get();
+    }
+    {
+      std::lock_guard<std::mutex> lk(payload_queue->mu_);
+      auto multiplier = (model_instance == nullptr)
```

Comment:
This `auto` is probably ok since it should get typed to `size_t` rather than an `int`. My style is to be explicit in these situations to tell the reader of the code we do not want integer overflow in the preceding lines. 

I will preface this next comment with, "I don't know the implicit casting rules when using the `<` operator." This has the very small potential to be a pain in line 224 if multiplier is typed to `int` and `(2 * multiplier)` gets overflowed. Again, this shouldn't be a problem since we don't expect `multiplier` to be greater than 2^32-1.